Base/usr/share/man/man2/futex.md

```diff
@@ -0,0 +1,191 @@
+at all in the common uncontended case, avoiding the cost of making a syscall
+completely.
+
+*A futex* is a single 32-bit integer cell located anywhere in the address space
+of a process (identified by its address), as well as an associated kernel-side
+queue of waiting threads. The kernel-side resources associated with a futex are
+created and destroyed implicitly when a futex is used; in other words, any
+32-bit integer can be used as a futex without any specific setup, and a futex
+on which no threads are waiting is no different to any other integer. The
+kernel does not assign any meaning to the value of the futex integer; it is up
+to userspace to make use of the value for its own logic.
+
+The `futex()` API provides a number of *operations*, the most basic ones being
+_waiting_ and _waking_:
+
+* `FUTEX_WAKE` / `futex_wake()`: wake up to `count` threads waiting on the
+  futex (in the raw `futex()` syscall, `count` is passed as the `value`
+  argument). The two most common values for `count` are 1 (wake a single
+  thread) and `UINT32_MAX` (wake all threads).
+* `FUTEX_WAIT` / `futex_wait()`: wait on the futex, but only if the current
+  value of the futex integer matches the specified `value`. The value
+  comparison and blocking is done atomically: if another thread changes the
+  value before the calling thread starts waiting, the calling thread will not
+  begin waiting at all, and the `futex_wait()` call will return `EAGAIN`
+  immediately. A waiting thread may wake up spuriously, without a matching call
+  to `futex_wake()`.
+* `FUTEX_WAKE_BITSET`: like `FUTEX_WAKE`, but only consider waiting threads
+  that have specified a matching bitset (passed in `value3`). Two bitsets match
+  if their *bitwise and* is non-zero. A thread that has not specified a bitset
+  is treated as having a bitset with all bits set (`FUTEX_BITSET_MATCH_ANY`,
+  equal to `0xffffffff`).
+* `FUTEX_WAIT_BITSET`: like `FUTEX_WAIT`, but the thread will only get woken by
+   wake operations specifying a matching bitset.
+* `FUTEX_REQUEUE`: wake up to `value` threads waiting on the futex, and requeue
+  up to `value2` (passed instead of the `timeout` argument) of the remaining
+  waiting threads to wait on another futex specified by `userspace_address2`,
+  without waking them up. Waking and requeueing threads is done atomically.
+
+  Requeueing threads without waking them up is useful to avoid "thundering
+  herd" issues with synchronization primitives like condition variables, where
+  multiple threads may wait for an event, but an event can only be handled by a
+  single thread at a time.
+* `FUTEX_CMP_REQUEUE`: like `FUTEX_REQUEUE`, but only if the current value of
+  the futex integer matches the specified `value3`. The value comparison,
+  waking and requeueing threads are all done atomically.
+* `FUTEX_WAKE_OP`: modify the value of the futex specified by
+  `userspace_address2`, wake up to `value` threads waiting on the futex, and
```

Comment:
Should this not read `userspace_address`? I.e. modification on the first futex, but wakeup on both first and second futex.

Reply:
No, according to both the Linux man page and the Serenity implementation, it modifies the value of the second futex.

https://github.com/SerenityOS/serenity/blob/65232b6681f33fd42508adf167100c155559274d/Kernel/Syscalls/futex.cpp#L257-L275